doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+  and handle them
+- Transparent interface for storing configuration parameters in non-volatile
+  storage devices
+
+## Status
+This document is currently under open discussion. This document is a product of
+[Configuration Task
+Force](https://github.com/RIOT-OS/RIOT/wiki/Configuration-Task-Force-(CTF)), and
+aims to describe the architecture of a runtime configuration system. The content
+of this document is licensed with a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) terminology
+and the following acronyms and definitions:
+
+### Acronyms
+- RDM: RIOT Developer Memo
+- CTF: RIOT Configuration Task Force
+- RCS: Runtime Configuration System
+- RH: Registry Handler
+- SF: Storage Facility
+
+### Definitions
+- Configuration group: A set of key-value configurations under the same name
+  scheme and scope. E.g `LIGHT_SENSOR_THRESHOLD` and `TRANSMISSION_PERIOD`
+configuration parameters can be contained in an _Application_ configuration
+group, as well as `IEEE802154_CHANNEL` and `IEEE802154_TX_POWER` in a
+_IEEE802.15.4 Radio_ configuration group. Within RIOT, a Configuration Group
+is represented by a Registry Handler.
+
+- Registry Handler: A descriptor that acts as an interface between the RIOT
+Registry and a module that exposes configurations. It provides a common
+interface to get, set, apply and export configurations.
+
+- Storage Facility: A descriptor that acts as an interface between the RIOT
+Registry and a non-volatile storage device. It provides a common interface for
+load/store key-value data from storage devices that might have different
+data representations.
+
+# 1. Introduction
+This document specifies the proposed architecture by the
+Configuration Task Force (CFT) to implement a secure and reliable Runtime
+Configuration System (RCS) focusing on modularity, reuse of existing
+technologies in RIOT (network stack, storage interface) and standard compliance.
+
+# 2. Architecture
+The proposed RCS architecture is formed by a
+[configuration manager](4-configuration-manager) and the
+[RIOT Registry](3-the-riot-registry) and its sub-components:
+one or more [Registry Handlers](4-registry-handlers) and one or more
+[storage facilities](5-storage-facilities). 
+
+A RIOT Application may interact with the Configuration Manager in order to
+modify access control rules or enable different communication interfaces.
+Also, it may interact with the RIOT Registry directly if it's
+required to load/store a persistent configuration.
+
+![](./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/architecture.svg "Runtime Configuration Architecture")
+
+# 3. The RIOT Registry
+The RIOT Registry is a module for interacting with 
+**persistent key-value configurations**. It's heavily inspired in the
+[Mynewt Config subsystem](https://mynewt.apache.org/latest/os/modules/config/config.html)
```

Comment:
Ok, thanks - would it be worth saying all this in the text? To convince the reader that the homework has been done and this is the best solution.

Reply:
Any thoughts?

Reply:
> To convince the reader that the homework has been done and this is the best solution.

Why do you consider this to be necessary? Why do you think is necessary to convince the reader this is the best solution?

It's sort of expected the RDM will consider the best efforts solution, right? (It's hard to confirm this is THE BEST solution but it's the best we could do)

Reply:
> You could include a reference to Zephyr as well?

Zephyr configuration system is based on the same Mynewt scheme. I would only add a reference to that (since it doesn't make sense IMO to add the same scheme twice)

> I raised a question, as a point of discussion - happy with however it's resolved. Could you please wait until review points have been resolved via discussion with the reviewer before resolving conversations?

Just to be clear, I just resolved a conversation about a diagram **that is going to be changed**. The way the architecture is shown will be changed in any case, so that was resolved to me. We will upload the new picture in a few minutes

Reply:
>  We will upload the new picture in a few minutes

Changed